Python/Product/PythonTools/PythonTools/Intellisense/TaskProvider.cs

```diff
@@ -446,44 +481,59 @@ abstract class TaskProvider : IVsTaskProvider, IDisposable {
-                    } else {
-                        // Apply the message to our collection
-                        changed |= msg.Apply(_items, _itemsLock);
-                    }
+                    // Reset event and handle all messages in the queue
+                    _workerQueueChanged.Reset();
+
+                    while (!cts.IsCancellationRequested) {
+                        WorkerMessage msg;
+                        lock (_workerQueue) {
+                            if (_workerQueue.Count == 0) {
+                                break;
+                            }
+                            msg = _workerQueue.Dequeue();
+                        }
+
+                        if (msg is WorkerMessage.AbortMessage) {
+                            // Ignore flush messages and cancel the loop
+                            flushMessages.Clear();
+                            cts.Cancel();
+                            break;
+                        } else if (msg is WorkerMessage.FlushMessage) {
+                            // Keep flush messages until we've exited the loop
+                            flushMessages.Enqueue(msg);
+                        } else {
+                            // Apply the message to our collection
+                            changed |= msg.Apply(_items, _itemsLock);
+                        }
 
-                    // Every second, we want to force another update
-                    if (changed) {
-                        var currentTime = DateTime.Now;
-                        if ((currentTime - lastUpdateTime).TotalMilliseconds > 1000) {
-                            Refresh();
-                            lastUpdateTime = currentTime;
-                            changed = false;
+                        // Every second, we want to force another update
+                        if (changed) {
+                            var currentTime = DateTime.Now;
+                            if ((currentTime - lastUpdateTime).TotalMilliseconds > 1000) {
+                                Refresh();
+                                lastUpdateTime = currentTime;
+                                changed = false;
+                            }
                         }
                     }
 
+                    if (cts.IsCancellationRequested) {
+                        break;
+                    }
+
                     // Reset the timeout back to 1 second
                     cts.CancelAfter(1000);
                 }
             } catch (OperationCanceledException) {
-                // Expected when the timeout expires
-            } catch (ObjectDisposedException ex) {
-                // We have been disposed.
-                Debug.Assert(
-                    ex.ObjectName == "BlockingCollection",
-                    "Handled ObjectDisposedException for the wrong type"
-                );
-                return;
+                // Occurs if we accidentally call Apply() on 

```

Comment:
AbortMessage. I was halfway through the comment and decided to rename the message, then never finished the comment.